### PR TITLE
perf: optimize Arrow JSON serialization hot path

### PR DESCRIPTION
**Pull request by strawgate**
Introduces a batch-level JSON serialization path with buffer pre-reservation and rewrites the inner string escape scan with SWAR bit tricks for faster control character detection.

- Add `write_batch_json_resolved` for batch-level serialization with pre-reserved buffers
- Replace per-row loops in `JsonLinesSink` and `StdoutSink` with single batch calls
- Accelerate `first_escape_pos` with 8-byte SWAR processing
- Add no-escape fast path in `write_json_string` to skip allocation when safe
- Replace `format_args!` with a static hex table and `extend_from_slice` for control characters

> [!NOTE]
> ### Optimize Arrow JSON serialization by replacing per-row loops with batch-level NDJSON writes
> - Introduces `write_batch_json_resolved` in [`row_json.rs`](https://github.com/strawgate/fastforward/pull/2727/files#diff-9f738244fd6e9e022e6de11f78d320f4377015d04bbb4103883cba4b8cb8f67f), which serializes an entire `RecordBatch` as NDJSON in one pass with pre-reserved buffer capacity.
> - Rewrites `write_escaped_json_string` to use a fast path for strings requiring no escaping and a SWAR-based 8-byte chunk scan for control characters, quotes, and backslashes, replacing `format_args!` with a fixed hex table.
> - Updates `JsonLinesSink` and `StdoutSink` to call `write_batch_json_resolved` instead of per-row `write_row_json_resolved` loops.
> - Adds benchmarks for `write_batch_json_resolved` on narrow and wide schemas in [`json_lines.rs`](https://github.com/strawgate/fastforward/pull/2727/files#diff-69b195325798b5e628e6732d232b1f0770240d6c034dd1b8a78496f3326647ec).
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 5869d66.</sup>
> 
>

### CODE REVIEW

**Comment by strawgate** [comment]
## 🚀 PR Rocket · Auto-pilot ✅ ON

Status: **Needs Human**
Last action: Agent run failed with exit code 137

PR Rocket is actively managing this PR. It will follow the conversation,
fix CI, resolve conflicts, keep the branch updated, and make changes quietly.


<details><summary>Recent activity</summary>

- Agent run failed with exit code 137
- Agent run failed with exit code 1
- ⚡ Review: The diff only contains GitHub webhook event JSON files — no source code changes
- ⚡ Preview title and body, labels (performance, enhancement, refactor): perf: optimize Arrow JSON serialization hot path
- ⚡ Preview title and body, labels (performance, enhancement): perf: optimize Arrow JSON serialization hot path
- Pushed changes to the branch
- ⚡ Review: Solid performance-focused PR with meaningful hot-path optimizations. The SWAR es
- ⚡ Preview title and body, labels (performance, enhancement): Optimize Arrow JSON serialization with batch writer and SWAR escaping

</details>

### Needs Human Input
Agent run failed (exit 137)

PR Rocket could not complete this pass because the sandboxed agent exited with code 137. Review the sandbox logs and retry.

---
`/rocket disable auto-pilot`

**Comment by strawgate** [comment]
### ⚡ Quick Review

Solid performance-focused PR with meaningful hot-path optimizations. The SWAR escape scanning and batch-level pre-reservation are well-targeted for the JSON serialization bottleneck. Low risk overall — logic is straightforward and preserves existing behavior.

- The SWAR implementation in `first_escape_pos` looks correct: the `unsafe` cast is properly guarded by the loop invariant `i + 8 <= len`, and the three-mask detection (control chars, quote, backslash) uses the standard zero-byte trick reliably. The trailing-zeros-to-byte-index conversion assumes little-endian, which matches all target platforms for this crate.
- `write_batch_json_resolved` uses a reasonable heuristic (`40 + cols.len() * 30` bytes/row) but over-reservation won't cause correctness issues — only minor memory overhead.
- The slow path in `write_json_string` now avoids `format_args!` for control chars using a static hex lookup table — this is a good micro-optimization.
- The benchmark addition for `write_batch_json_resolved` is good. The PR would benefit from a quick sanity check that `write_batch_json_resolved` produces byte-identical output to the prior per-row loop (could be added as a property test or compile-time snapshot).
- Minor: the diff shows `write_row_json_resolved` gaining a trailing newline parameter (`true`) but I don't see that signature change in the `row_json.rs` diff — worth verifying the call sites in `file_output_profile.rs` are consistent with the function definition.

---
*Fast review by PR Rocket — the full agent will follow.*

**Comment by strawgate** [comment]
### ⚡ Quick Review

The diff only contains GitHub webhook event JSON files — no source code changes are visible. Please provide the actual Rust source diff for a proper review of the SWAR escape detection, batch serialization, and hex table escaping logic.

---
*Fast review by PR Rocket — the full agent will follow.*

**Comment by strawgate** [comment]
PR Rocket needs human help: **Agent run failed (exit 137)**

PR Rocket could not complete this pass because the sandboxed agent exited with code 137. Review the sandbox logs and retry.

**Comment by strawgate** [comment]
Performance improvements merged

File: crates/ffwd-output/src/row_json.rs [hunk 1]
```diff
@@ -527,7 +615,32 @@ pub fn write_row_json_resolved(
...
+    if num_rows == 0 {
+        return Ok(0);
+    }
+    // Estimate ~200 bytes per row for narrow, ~600 for wide schemas.
+    // Use column count as a heuristic: 5 cols ~ 200B, 20 cols ~ 600B.
+    let est_bytes_per_row = 40 + cols.len() * 30;
+    out.reserve(num_rows * est_bytes_per_row);
+
+    for row in 0..num_rows {
+        write_row_json_resolved(row, cols, out, true)?;
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/row_json.rs:632]: Added a row-limit guard for `write_batch_json_resolved` so safe callers cannot pass a larger `num_rows` than the resolved columns support, and covered it with `write_batch_json_resolved_rejects_oversized_num_rows`.

File: crates/ffwd-output/src/row_json.rs [hunk 2]
```diff
@@ -68,59 +68,147 @@ const NEEDS_ESCAPE: [bool; 256] = {
...
+        // Zero-byte detection: (v - 0x0101...) & !v & 0x8080... != 0
+
+        let lo_mask = 0x8080_8080_8080_8080u64;
+        let ones = 0x0101_0101_0101_0101u64;
+
+        // Check for control chars (< 0x20):
+        // For unsigned subtraction: if any byte < 0x20, then (byte - 0x20) wraps.
+        // We detect this via: has_value_lt(chunk, 0x20)
+        // = ((chunk - ones * 0x20) & !chunk & lo_mask) != 0
+        let ctrl = (chunk.wrapping_sub(ones.wrapping_mul(0x20))) & !chunk & lo_mask;
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/row_json.rs]: Switched the control-byte detection to the top-3-bits-zero SWAR form and added randomized plus edge-case regression tests that compare `first_escape_pos` against the reference byte-scan implementation.

File: crates/ffwd-output/src/row_json.rs [hunk 3]
```diff
@@ -527,7 +615,32 @@ pub fn write_row_json_resolved(
...
+    num_rows: usize,
+    out: &mut Vec<u8>,
+) -> io::Result<usize> {
+    if num_rows == 0 {
+        return Ok(0);
+    }
+    // Estimate ~200 bytes per row for narrow, ~600 for wide schemas.
+    // Use column count as a heuristic: 5 cols ~ 200B, 20 cols ~ 600B.
+    let est_bytes_per_row = 40 + cols.len() * 30;
+    out.reserve(num_rows * est_bytes_per_row);
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/row_json.rs:629]: Changed the heuristic preallocation to checked math plus `try_reserve`, and capped the speculative preallocation so oversized batches fail gracefully instead of risking a large reserve.

File: crates/ffwd-output/src/row_json.rs [hunk 4]
```diff
@@ -527,7 +615,32 @@ pub fn write_row_json_resolved(
     Ok(())
 }
 
-/// Write a single row as a JSON object into `out`.
+/// Write an entire batch as newline-delimited JSON using pre-resolved columns.
+///
+/// This is the highest-performance batch serialization path. Compared to calling
+/// [`write_row_json_resolved`] in a loop, this function:
+/// - Pre-reserves the output buffer based on batch size estimation
+/// - Amortizes the column resolution cost across all rows
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/row_json.rs:615]: Updated the doc comment to describe the helper accurately as a pre-reserving convenience loop over already-resolved columns.

**Comment by strawgate** [comment]
Reopened — PR needs rebasing on main. The merge conflicts are due to base commit being outdated (mergeStateStatus: DIRTY). Please run `git rebase main` on your branch.
